common/src/main/java/org/opensearch/sql/common/grok/Discovery.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.common.grok;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.StringUtils;
+
+
+/**
+ * {@code Discovery} try to find the best pattern for the given string.
+ *
+ * @since 0.0.2
+ */
+public class Discovery {
+
+  private Grok grok;
+
+  /**
+   * Create a new {@code Discovery} object.
+   *
+   * @param grok instance of grok
+   */
+  public Discovery(Grok grok) {
+    this.grok = grok;
+  }
+
+  /**
+   * Sort by regex complexity.
+   *
+   * @param groks Map of the pattern name and grok instance
+   * @return the map sorted by grok pattern complexity
+   */
+  private Map<String, Grok> sort(Map<String, Grok> groks) {
+
+    List<Grok> groky = new ArrayList<Grok>(groks.values());
+    Map<String, Grok> grokMap = new LinkedHashMap<String, Grok>();
+    Collections.sort(groky, new Comparator<Grok>() {
+      public int compare(Grok g1, Grok g2) {
+        return (this.complexity(g1.getNamedRegex()) < this.complexity(g2.getNamedRegex())) ? 1 : 0;
+      }
+
+      private int complexity(String expandedPattern) {
+        int score = 0;
+        score += expandedPattern.split("\\Q" + "|" + "\\E", -1).length - 1;
+        score += expandedPattern.length();
+        return score;
+      }
+    });
+
+    for (Grok grok : groky) {
+      grokMap.put(grok.getSaved_pattern(), grok);
+    }
+    return grokMap;
+
+  }
+
+  /**
+   * Determinate the complexity of the pattern.
+   *
+   * @param expandedPattern regex string
+   * @return the complexity of the regex
+   */
+  private int complexity(String expandedPattern) {
+    int score = 0;
+
+    score += expandedPattern.split("\\Q" + "|" + "\\E", -1).length - 1;
+    score += expandedPattern.length();
+
+    return score;
+  }
+
+  /**
+   * Find a pattern from a log.
+   *
+   * @param text witch is the representation of your single
+   * @return Grok pattern %{Foo}...
+   */
+  public String discover(String text) {
+    if (text == null) {
+      return "";
+    }
+
+    Map<String, Grok> groks = new TreeMap<String, Grok>();
+    Map<String, String> grokPatterns = grok.getPatterns();
+    // Boolean done = false;
+    String texte = text;
+    GrokCompiler compiler = GrokCompiler.newInstance();
+    compiler.register(grokPatterns);
+
+    // Compile the pattern
+    for (Entry<String, String> stringStringEntry : grokPatterns.entrySet()) {
+      @SuppressWarnings("rawtypes")
+      Entry pairs = (Entry) stringStringEntry;
+      String key = pairs.getKey().toString();
```

Comment:
Why is this necessary?

Reply:
[java-grok](https://github.com/thekrakken/java-grok) is an independent library that's used for the PPL's `grok` command, but in the latest release `Grok` is not serializable (and SQL grok expression cannot be pushed down). After discussing with peng I'm tentatively moving the source code into sql.common module.

And for this line, yes it's not required. Historically it was an iterator `Map.Entry pairs = (Map.Entry) it.next();`, and then was refactored to for each loop but that line was not removed. Actually the Discovery part is not used, will remove.